packages/react-utilities/src/compose/types.ts

```diff
@@ -48,3 +49,54 @@ export type SlotProps<TSlots extends BaseSlots, TProps, TRootProps extends React
 } & {
   root: TRootProps;
 };
+
+/**
+ * Helper type to convert the given props of type ShorthandProps<...> into ObjectShorthandProps<...>
+ */
+export type ResolvedShorthandProps<Props, ShorthandPropNames extends keyof Props> = Omit<Props, ShorthandPropNames> &
+  { [P in ShorthandPropNames]: Props[P] extends ShorthandProps<infer T> ? ObjectShorthandProps<T> : Props[P] };
+
+/**
+ * Helper type to mark the given props as required.
+ * Similar to Required<T> except it only requires a subset of the props.
+ */
+export type RequiredProps<Props, RequiredProps extends keyof Props> = Omit<Props, RequiredProps> &
+  { [P in RequiredProps]-?: Props[P] };
+
+/**
+ * Converts a components Props type to a State type:
+ * * Adds the 'ref' and 'as' types as required values
+ * * Ensures the specified ShorthandProps are of type ObjectShorthandProps<T>
+ * * Marks the given DefaultedProps as required (-?)
+ *
+ * Example usage:
+ * ```typescript
+ * export type ExampleState = ComponentState<
+ *    ExampleProps,
+ *    { ShorthandProps: 'icon' | 'text' },
+ *    { DefaultedProps: 'color' | 'size' }
+ *  >;
+ * ```
+ */
+export type ComponentState<
+  Props extends ComponentProps,
+  ShorthandProps extends { ShorthandProps: keyof Props },
+  DefaultedProps extends { DefaultedProps: keyof Props } = { DefaultedProps: never }
+> = RequiredProps<
+  ResolvedShorthandProps<Props, ShorthandProps['ShorthandProps']>,
```

Comment:
The usage won't actually be that bad, but it can be confusing when looking at these types and wondering why `ShorthandProps['ShorthandProps']` is being used.

I don't have a better solution though and the usage of this looks great 🤷

Reply:
Removed the nesting and updated it to look more like your suggestion.

Reply:
I would just briefly nit and suggest to perhaps keep the `T` prefix ? it seems that it's a standard thing for generics when I read codebases... but perhaps I'm wrong it's not  a blocker for me